lib/rex/io/stream.rb

```diff
@@ -61,7 +61,7 @@ def write(buf, opts = {})
             next
           end
           data = buf[total_sent, block_size]
-          sent = fd.write_nonblock( data )
+          sent = fd.syswrite( data )
```

Comment:
One piece of information missing from the description:

This line _used_ to be `syswrite`, but was changed to `write_nonblock` a few years ago:

> Change stream writes to be non-blocking, this has been pretty heavily tested over the last couple days, and should preserve the original behavior as well. This fixes a deadlock in Ruby in certain conditions
>
> git-svn-id: file:///home/svn/framework3/trunk@10306 4d416f70-5f16-0410-b530-b9f4589650da
> ```diff
> - 				sent = fd.syswrite( data )
> +				sent = fd.write_nonblock( data )
> ```
>https://github.com/rapid7/rex-core/commit/7a7170baada7d4ff97e97bb75ddad6c69e438a25

So there's potentially more gremlins beyond this one line of code. I really wish there was more detail on the original bug fix, so we could replicate and fix those issues too. From the description alone, we weren't sure if it was the Ruby VM itself locking up, or Metasploit's Ruby code. If it was the Ruby VM, potentially it's fixed now. The only fear is flip/flopping on this line of code around to fix one problem, and accidentally re-introduce the original problems unintentionally 😄